### PR TITLE
[Core]Support branch

### PR DESCRIPTION
**Pull request by TaoZex**
### Purpose

Introduce branch for paimon
Linked issue: https://github.com/apache/incubator-paimon/issues/2148
deisgn：https://cwiki.apache.org/confluence/display/PAIMON/PIP-9%3A+Support+Branch
task：https://github.com/apache/incubator-paimon/issues/1795

### Tests

FileStoreTableTestBase

### CODE REVIEW

**Comment by TaoZex** [comment]
The functions of merge branch and replace branch are complete, but I need to implement writing data to branch to test the usability of these functions. A new pr commit will be created after I finish writing data to the branch.

File: paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+    public Path branchSnapshotPath(String branchName, long snapshotId) {
+        return new Path(getBranchPath(branchName) + "/snapshot/" + SNAPSHOT_PREFIX + snapshotId);
+    }
+
+    public Path schemaPath(long schemaId) {
+        return new Path(tablePath + "/schema/" + SCHEMA_PREFIX + schemaId);
+    }
+
+    public Path branchSchemaPath(String branchName, long schemaId) {
```
Comment (FangYongs) [inline paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java:84]: We should not build tag, snapshot and schema path in `BranchManager`

File: paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            }
+            return false;
+        } catch (IOException e) {
+            throw new RuntimeException(
+                    String.format("Failed to determine if path '%s' exists.", path), e);
+        }
+    }
+
+    /** Check if a tag exists. */
+    public boolean tagExists(String tagName) {
```
Comment (FangYongs) [inline paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java:144]: We should not check if a tag exist in `BranchManager`

File: paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    /** Check if a branch exists. */
+    public boolean branchExists(String branchName) {
+        Path branchPath = branchPath(branchName);
+        return fileExists(branchPath);
+    }
+
+    /** Get the tagged snapshot by name. */
+    public Snapshot taggedSnapshot(String tagName) {
```
Comment (FangYongs) [inline paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java:156]: We should not get `Snapshot` for given tag name in `BranchManager` like this

File: paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java [hunk 4]
```diff
@@ -50,7 +50,7 @@ public class SnapshotManager implements Serializable {
 
     private static final long serialVersionUID = 1L;
 
-    private static final String SNAPSHOT_PREFIX = "snapshot-";
+    protected static final String SNAPSHOT_PREFIX = "snapshot-";
```
Comment (FangYongs) [inline paimon-core/src/main/java/org/apache/paimon/utils/SnapshotManager.java:53]: The internal prefix should not be accessed by `BranchManager`

File: paimon-core/src/main/java/org/apache/paimon/utils/TagManager.java [hunk 5]
```diff
@@ -46,7 +46,7 @@ public class TagManager {
 
     private static final Logger LOG = LoggerFactory.getLogger(TagManager.class);
 
-    private static final String TAG_PREFIX = "tag-";
+    protected static final String TAG_PREFIX = "tag-";
```
Comment (FangYongs) [inline paimon-core/src/main/java/org/apache/paimon/utils/TagManager.java:49]: As above

File: paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        checkArgument(
+                !branchName.chars().allMatch(Character::isDigit),
+                "Branch name cannot be pure numeric string but is '%s'.",
+                branchName);
+
+        Snapshot snapshot = taggedSnapshot(tagName);
+
+        try {
+            // Copy the corresponding tag, snapshot and schema files into the branch directory
+            fileIO.copyFileUtf8(tagPath(tagName), branchTagPath(branchName, tagName));
```
Comment (FangYongs) [inline paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java:101]: Why should we copy tag file to branch directory?

**Comment by FangYongs** [comment]
Hi @TaoZex , are you still working on this PR?

**Comment by TaoZex** [comment]
> Hi @TaoZex , are you still working on this PR?

Thanks for your review, I am still working on this pr, and I will improve it according to your suggestions this week. @FangYongs

File: paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java (hunk 1 above)
Comment (TaoZex) [reply paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java]: done. Thanks for your review @FangYongs

File: paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java (hunk 6 above)
Comment (TaoZex) [reply paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java]: From design:
image

In addition, I think it makes sense to copy tag. The new branch will make it easier for users to use tag. What do you think about it？

**Comment by FangYongs** [comment]
@TaoZex  I found that this PR has added "merge" relevant codes, involving modifications to the file store without tests. Can we follow the original PR implementation to only introduce the `BranchManager` to support `create` and `delete` branch but not involve modifications to the file store? I think this can reduce the risk and merge this PR as soon as possible.

**Comment by TaoZex** [comment]
@FangYongs Thanks for your advice. I will make changes according to your suggestions, and submit a new pr about merge and replace branch after this pr is merged(with file store tests).

**Pull request closed without merging**